# Essential Exploratory Data Analysis (EDA) Methods and Identifying Missing Values with Pandas

This notebook introduces key Pandas methods for identifying missing values and performing some essential exploratory data analysis (EDA) for basic information on a dataset.

## Setup

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

## Example DataFrame with Missing Values
We'll look at a Formula 1 dataset of races and drivers. Note the `NaN` values where data is missing.

In [2]:
df  = pd.DataFrame({
    "race_1": [25, np.nan, 18, 1],
    "race_2": [25, 12, np.nan, np.nan],
    "race_3": [25, 12, 18, np.nan]
}, index =  ["Max Verstappen", "Lewis Hamilton", "Fernando Alonso", "Lando Norris"])

df

,race_1,race_2,race_3
Max Verstappen,25.0,25.0,25.0
Lewis Hamilton,NaN,12.0,12.0
Fernando Alonso,18.0,NaN,18.0
Lando Norris,1.0,NaN,NaN


## Identifying Missing Values

`df.col.isna()` or `df.col.isnull()` can be used to detect missing values. Both return a Boolean Series where:

- True indicates a missing value (`NaN`, `None`, or `NaT` for "not a time")

- False indicates a non-missing value

In [3]:
df['race_3'].isna()

Max Verstappen     False
Lewis Hamilton     False
Fernando Alonso    False
Lando Norris        True
Name: race_3, dtype: bool

### Total Missing Values
You can quickly calculate the number of missing values in a column by running `.sum()` on `.isna()`.

In [4]:
df['race_3'].isna().sum()

1

On certain versions of Pandas, you might see values returned like `np.int64(1)` instead of just `1`. That's fine: it's the data type as well as the value, which is whatever number is inside the parentheses.

You can also call `.isna().sum()` on an entire dataset, which will provide the number of missing values for every column.

In [3]:
df.isna().sum() # returns total number of missing values in each column of the DataFrame.

race_1    1
race_2    2
race_3    1
dtype: int64

### Proportion of Missing Values

In [6]:
df.isna().mean() # returns mean number of missing values in each column of the DataFrame.

race_1    0.25
race_2    0.50
race_3    0.25
dtype: float64

## Dropping Missing Values

- Drops rows that contain at least one missing value (NaN, None, NaT).

- Returns a new DataFrame (it doesn’t change the original dataframe unless you use inplace=True).

Doing this is sometimes referred to as keeping only the **complete cases**, i.e. rows where all columns have a value.



In [4]:
complete_cases = df.dropna()
complete_cases

,race_1,race_2,race_3
Max Verstappen,25.0,25.0,25.0


### Drop rows vs drop columns


axis=0 (default): drop rows

axis=1: drop columns

In the below example, each column has at least one missing value, so all columns are dropped!

In [5]:
df.dropna(axis=1)

""
Max Verstappen
Lewis Hamilton
Fernando Alonso
Lando Norris


### Drop rows with missing values in particular columns
Sometimes you don't want to drop rows that have any missing value, but only rows with missing values in particular columns that may be important. In that case, you can use `.dropna(subset=[columns])`.

In [9]:
df.dropna(subset=['race_2'])

,race_1,race_2,race_3
Max Verstappen,25.0,25.0,25.0
Lewis Hamilton,NaN,12.0,12.0


The `how` parameter of `dropna` determines the strictness:

how="any" (default): drop row/column if any value is missing

how="all": drop row/column only if all values are missing

In [10]:
df.dropna(how="all") # Doesn't delete anything since every row has one non-null value

,race_1,race_2,race_3
Max Verstappen,25.0,25.0,25.0
Lewis Hamilton,NaN,12.0,12.0
Fernando Alonso,18.0,NaN,18.0
Lando Norris,1.0,NaN,NaN


## Essential EDA Methods
Here are some basic exploratory analyses that you'll want to know for almost all datasets you'll be working with. Before you get into further analysis, it's good to know:
* How many rows are in the dataset
* What columns are in the dataset
* What the data type of each column is
* The number of unique values per column
* The number of missing values per column

This information will often affect how you prepare the data for later visualization or analysis. For example, if there are lots of missing values in an important column, you might want to remove those missing values before analyzing the dataset. Or if there are just a handful of values for a continuous (numeric) column, you may want treat it as a categorical column instead.

### `.info()` for Quick DataFrame Information
The `df.info()` function provides much of this information at a glance, including the number of rows (entries), names of columns and their datatypes.

In [11]:
df.info() # Returns information about a DataFrame including the index dtype and columns, non-NA values and memory usage.

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, Max Verstappen to Lando Norris
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   race_1  3 non-null      float64
 1   race_2  2 non-null      float64
 2   race_3  3 non-null      float64
dtypes: float64(3)
memory usage: 300.0+ bytes


`.info()` provides the number of non-null values for every column, which can lead to some mental gymnastics to see how many missing values there are per column, specifically subtracting the number of non-null values from the total number of rows. As discussed above, it's often nice to directly get the number of missing values for each column.

In [6]:
df.isna().sum()

race_1    1
race_2    2
race_3    1
dtype: int64

### Column Names and Data Types
It's often useful to get a list of the column names directly. `df.columns` is your go-to tool for that.

In [ ]:
df.columns # Returns the column labels of the DataFrame.

If you want that in list form, you can run:

In [ ]:
df.columns.tolist()

Other times you might want to get column data types directly.

In [ ]:
df.dtypes

### Number of Unique Values per Column

In [ ]:
df.nunique() #Returns the number of unique values per column.

By default, `NaN` is not included as a unique value. But you can include it by passing `dropna=False` to `nunique()`.

In [ ]:
df.nunique(dropna=False)

## Descriptive Statistics

Outside of the basic essential EDA, Pandas offers some handy functions for automatically getting information about the distribution of values within continuous (numeric).

`describe()` generates summary statistics for each continuous column in the DataFrame:
- count – number of non-missing values

- mean – average value

- std – standard deviation (measure of spread)

- min – minimum value

- 25% – first quartile (Q1) value

- 50% – median

- 75% – third quartile (Q3) value

- max – maximum value

In [ ]:
df.describe()

### Descriptive Statistics for Categorical Columns
Can you run `.describe()` for categorical columns? Yes, with the `include='object'` argument. It provides a different set of information more relevant to categorical columns:
* `count`: the count of non-null rows in the column
* `unique`: the number of unique values in the column
* `top`: the most frequent value
* `freq`: the most common value's frequency

Let's look at how this works with a dataframe with a categorical column.

In [7]:
df_2 = pd.DataFrame({
    "race_1": [25, np.nan, 18, 1],
    "race_2": [25, 12, np.nan, np.nan],
    "race_3": [25, 12, 18, np.nan],
    "teams" : ["Red Bull", "Mercedes", "Ferrari", "McLaren"]
}
, index =  ["Max Verstappen", "Lewis Hamilton", "Fernando Alonso", "Lando Norris"])

df_2.describe(include='object')

,teams
count,4
unique,4
top,Red Bull
freq,1


Each value just had a count of 1 in the `teams` column, so Pandas chose the first one (Red Bull) as the top.

## Realistic Example: Titanic Dataset
Let's apply these techniques to a new dataset of Titanic passengers.

In [8]:
titanic = sns.load_dataset('titanic')
titanic.info()
titanic.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


We can see from `.info()` that there are 891 rows and 15 columns in the dataset. Column names and data types are also listed. From the first 5 rows and the Non-Null Count in `info()`, we can see that there are some missing (`NaN`) values. Let's see exactly how many.

In [9]:
titanic.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Digging a bit deeper, let's see how many unique values we have for each column. We can treat continuous variables as categorical if they have a small number of unique values.

In [10]:
titanic.nunique()

survived         2
pclass           3
sex              2
age             88
sibsp            7
parch            7
fare           248
embarked         3
class            3
who              3
adult_male       2
deck             7
embark_town      3
alive            2
alone            2
dtype: int64

As a reminder, `.value_counts()` provides a useful way to obtain the number of rows per each unique value. Here we'll apply it to the `pclass` column, which is the class of the passenger's ticket (first, second, or third).

In [11]:
titanic['pclass'].value_counts()

pclass
3    491
1    216
2    184
Name: count, dtype: int64

For continuous columns, we'll use .describe() to quickly obtain summary statistics. Note that these might not be as helpful for columns like `pclass` with only a few unique values, whether the passenger was in first, second, or third class. For those, it might be more helpful to know things like which is the most frequent value (as we investigated above).

In [ ]:
titanic.describe()

## Summary
- `isna()` and `isnull()` identify missing values. `isna().sum()` and `isna.mean()` quantify missing data
- `dropna()` returns complete cases
- Methods like `.info()`, `.nunique()`, and `.describe()` provide quick insights